swift/swift.go

```diff
@@ -188,3 +204,67 @@ func maybeNotFound(err error, format string, arg ...interface{}) error {
 	}
 	return errors.Newf(err, format, arg...)
 }
+
+type ohandler struct {
+	c             *Client
+	containerName string
+	objectName    string
+	length        int64
+	pos           int64
+}
+
+// Read reads up to len(b) bytes from the handler. It returns the number of bytes read and any error encountered. At end of stream, Read returns no EOF on the first EOF and EOF after that.
+func (o *ohandler) Read(p []byte) (n int, err error) {
+	if o.pos == o.length {
+		return 0, io.EOF
+	}
+	end := int64(len(p)) + o.pos
+	header := make(http.Header)
+	header.Add("Range", fmt.Sprintf("bytes=%d-%d", o.pos, end))
+	requestData := goosehttp.RequestData{
+		RespReader:     &emptyReadCloser,
+		ReqHeaders:     header,
+		ExpectedStatus: []int{http.StatusOK, http.StatusPartialContent},
+	}
+	err = o.c.touchObject(&requestData, client.GET, o.containerName, o.objectName)
+	if err != nil {
+		return
+	}
+	defer requestData.RespReader.Close()
+	readSize := 0
+	for n < len(p) {
+		readSize, err = requestData.RespReader.Read(p[n:])
+		n += readSize
+		o.pos += int64(readSize)
+		if err != nil {
+			if err == io.EOF || err == io.ErrUnexpectedEOF {
+				// Some code assumes certain semantics about when EOF is returned.
+				err = nil
+				// Set pos to length, this should be noop but its possible something lies.
+				o.pos = o.length
+			}
+			return
+		}
+	}
+	return
+}
+
+// Seek sets the offset for the next Read operation on the handler.
+func (o *ohandler) Seek(offset int64, whence int) (int64, error) {
```

Comment:
The contract for seeker suggests that

> Seeking to an offset before the start of the file is an error. Seeking to any positive offset is legal, but the behavior of subsequent I/O operations on the underlying object is implementation-dependent.

I think implementing that behavior would be worthwhile.